### DISCUSSION TITLE
Adding support for Z3's fp.to_ieee_bv (or workaround?)

### DISCUSSION THREAD

**Discussion by bavalpey** [cvc5/cvc5#9277] [Q&A]
To start, I am trying to use CVC to verify properties of a machine for which I have created a model in CVC. 
 The machine does not differentiate between the types of its variables, rather, the instructions interpret them as a specific type.  To this end, I have modeled the machine's registers as bitvectors.  I have written, in smtlib, a "library" of sorts, using smtlib's ``(define-fun)`` macro, for each of the instructions.  As some of the instructions interpret registers as floating point types, I need to be able to convert between bitvectors and floats.  Of course, there is support for converting from a bitvector with ``(_ to_fp eb sb)``, but no way to get back to the bitvector afterwards.

[SMTlib's FP Theory](http://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml) suggests the following:
```smt2
(declare-fun b () (_ BitVec m))
(assert (= ((_ to_fp eb sb) b) f))
```
Unfortunately, this only works for top level assertions. This prevents me (as far as I know) from being able to perform this kind of conversion in the body of my ``(define-fun)`` macro. 

Since Z3 supports fp.to_ieee_bv, I have been using it for my verification.  However, CVC performs substantially better on the verification tasks than Z3, in some cases taking less than a second on a script that takes Z3 more than 24 hours.  I would therefore really like to be able to use CVC for verification.
 
I saw in issue #5537 that one workaround is to use a forall, perhaps with a trigger/pattern.  I tried this below in a simple script to see if this might work for my use case.

```smt
(set-logic ALL)
(declare-fun my_fp_ieee_bv32 (Float32) (_ BitVec 32))
(assert (forall ((f Float32))
    (!
        (ite
            (fp.isNaN f)
            (= (my_fp_ieee_bv32 f) (ite (fp.isPositive f) #x7f100000 #xff100000))
            (= ((_ to_fp 8 24) (my_fp_ieee_bv32 f)) f))
        :pattern ((my_fp_ieee_bv32 f)))))
(declare-fun a () Float32)
(declare-fun b () Float32)
(assert (= a (fp #b0 #x7f #b10000000000000000000000)))
(assert (= b (fp #b0 #x7f #b10100000000000000000000)))
(assert (= (my_fp_ieee_bv32 (fp.add RNE a b)) (my_fp_ieee_bv32 (fp.add RNE b a))))

(check-sat)
```

The above script returns ``unknown`` (using cvc5 version 1.0.2 [git tag 1.0.2 branch HEAD]).  This isn't workable for me: the script must be able to return a response for such trivial operations. Since z3 supports fp.to_ieee_bv, where NaNs are simply uninterpreted, I can write the same script instead _defining_ ``my_fp_ieee_bv32`` (instead of declaring it), utilizing ``fp.to_ieee_bv``

For my use case, I need to be able to convert floating point types back to their bitvector counterpart.  If the floating point is NaN, then I don't care _which_ bitvector value is returned, only that the value is the same between multiple invocations on NaN values (in my example, +NaN and -NaN were different, although I don't even require this functionality).

Now, I was trying to determine how I might go about forking cvc5 and adding this functionality myself, but ran into trouble in doing so.  I tried to replicate what is being done for fp.to_real, since my function should have similar behavior (undefined for NaN values).  I looked through the source code and tried to make the appropriate modifications.  Unfortunately, I was unable to identify all of the places in the source code that needed to be changed in order to support this kind of instruction.

Specifically, I modified the following

- Adding a KIND_ENUM in ``src/api/cpp/cvc5.cpp`` - ``KIND_ENUM(FLOATINGPOINT_TO_IEEE_BV, internal::Kind::FLOATINGPOINT_TO_IEEE_BV)`` and ``{internal::Kind::FLOATINGPOINT_TO_IEEE_BV_FROM_FP_OP, FLOATINGPOINT_TO_IEEE_BV_FROM_FP}, {internal::Kind::FLOATINGPOINT_TO_IEEE_BV_FROM_FP, FLOATINGPOINT_TO_IEEE_BV_FROM_FP},`` in the same file

- Adding an operator and typerule in ``src/theory/fp/kinds`` ``operator FLOATINGPOINT_TO_IEEE_BV 1 "floating-point to IEEE-754 bit vector"`` ``typerule FLOATINGPOINT_TO_IEEE_BV ::cvc5::internal::theory::fp::FloatingPointToIEEEBVTypeRule``
- Adding a definition in ``src/api/cpp/cvc5_kind.h`` for ``FLOATINGPOINT_TO_IEEE_BV``.
- Adding a check for the rewrite in ``src/theory/fp/theory_fp.cpp`` (line 155): `` && k != kind::FLOATINGPOINT_TO_IEEE_BV``
- Adding to ``src/theory/fp/fp_expand_defs.cpp`` - ``d_toIEEEBVMap(u)`` on line 29, and a definition for ``Node FpExpandDefs::toIEEEBVUF(Node node)``.  This is the point at which I was unable to identify where I should further proceed.  I was unable to identify where ``d_toIEEEBVMap`` should be defined (like the ``d_toRealMap``). 

Any advice on the remaining steps that would be required to fully support this kind of instruction would be greatly appreciated.

**Comment by martin-cs** [comment]
HI @bavalpey 

Thanks for getting in touch and it is really impressive to see the work you have already put into this.

Perhaps it might help to understand why I didn't implement this.  The SMT-LIB floating-point theory works at "level 3" of representation of IEEE-754 so it does not work directly with bit-vectors.  This is intentional to allow solver algorithms based on reals and other non-BV techniques.  cvc5's theory solver is currently based on bit-blasting but that is a solver-specific design choice, not a consequence of the formalism.

An unfortunate consequence of this is that NaN really doesn't have a bit-pattern in the theory.  There are a (large) set of bit-vectors which convert to a NaN but we can't define a function for the reverse transformation because it really isn't a function.  Do you return an "Intel style" x86 positive NaN or an "ARM style" negative NaN?  Do you return a bit pattern that corresponds to a quiet or a signalling NaN?  Do you return a payload and if so what?  These may seem like trivial questions but hard-coding a solution means that /some/ of your users won't be able to replay the error traces generated by tools on their platform without additional, non-trivial, post processing.

The quantifier you give is probably the least hassle way of getting the functionality you want.  I would expect something like this to work but I know that triggers are *very* sensitive to syntactic structure.  Maybe @ajreynol can suggest some options for looking at why this doesn't work.

Anyway, if you do want to go this route, you will need to:

- Add a new kind in `src/theory/fp/kinds`.  I would recommend copying the handling of `to_real` rather than the conversion from bit-vectors as there is a certain amount of magic that is required to make polymorphic things work.  I would copy the split of to_real and to_real_total as functions should be total.
- You will need type rules for both kinds, see `src/theory/fptheory_fp_type_rules.{h,cpp}`.
- You will need to at least be able to rewrite total constants, see `src/theory/theory_fp_rewriter.{h,cpp}`.  IIRC `util/floatingpoint.h:FloatingPoint::pack()` should do the job.  Note that this will fix a particular value for NaN.  If you have an eye to efficiency, you might want to consider rewrites for BV -> FP -> BV (this one is non-trivial IIRC) and FP -> BV -> FP.
- As you noted you need to add to expandDefinition to create the uninterpreted function and change your kind for the total kind.
- Add parsing to `parser/smt2/smt2.cpp` and printing to `printer/smt2/smt2_printer.cpp`, these are really minimal changes.
- The most awkward part is going to be the bit-blasting.  For performance reasons, cvc5 does not convert each floating-point expression to a bit-vector expression; it converts whole predicates.  You will probably need to to catch the registration of the new total kind and in the word_blaster produce an auxiliary constraint which is a bit-vector equality something like: `total_expr(v, uf) == ITE(isNaN(v), mask(uf), pack(v))` where mask is a bit-vector expression which converts an arbitrary bit-vector to a NaN (this probably needs another auxiliary constraint) and pack is the function which packs a symbolic float into a symbolic bit vector.   This bit might be a bit tricky so I am happy to help if needed.

( Oh, as a minor aside, you are best off avoiding converting floats to/from bit-vectors between each instruction if you can help it.  The design of SymFPU (the bit blasting library used by cvc5 and bitwuzla for floating-point) assumes this is a rare occurrence and makes pack and unpack an expensive operation to make other things cheap. )

HTH

**Reply by bavalpey** [reply]
Wow thanks for the thorough response.

> An unfortunate consequence of this is that NaN really doesn't have a bit-pattern in the theory. There are a (large) set of bit-vectors which convert to a NaN but we can't define a function for the reverse transformation because it really isn't a function. 

Yes I absolutely agree.  I should reiterate that I want the behavior for NaNs to be the same for this function as it is for the to_real, where it is left uninterpreted, rather than deciding on a representation. 
In other words, I don't want my implementation to respond with a singular value for NaN.  I would like to leave it uninterpreted as it is for to_real.  For my use case, I would then just define my conversion that can handle NaN as follows:
```smt
(define-fun my_fp_ieee_bv32 ((a Float32)) (_ BitVec 32)
  (ite (fp.isNaN a)
       (ite (fp.isNegative a) bvNegNan bvPosNaN)
       (fp.to_ieee_be a)))
 ; bvPos/NegNaN can be whatever I want, doesn't matter.
; fp.to_ieee_bv is the name of the conversion
```


> You will need to at least be able to rewrite total constants ... Note that this will fix a particular value for NaN

I would like to not fix NaN values.  I'm away from my machine at the moment, but how does fp.to_real handle NaNs, given that they are supposed to be left uninterpreted? (Are they?) I would like my implementation to do the same thing and not return any value, perhaps leaving it as a lambda?

> Oh, as a minor aside, you are best off avoiding converting floats to/from bit-vectors between each instruction if you can help it

I thought about this, and at least for my use case plan on keeping things as floats for as long as possible.

For the remainder, I will likely have more questions once I begin working out your suggestions.  In any case, thanks again for your extremely detailed response.

**Comment by martin-cs** [comment]
Oh, I forgot, you also have to add your new kind to the equality engine in `theory_bv.cpp`.  If you search for where `kind::FLOATINGPOINT_TO_REAL_TOTAL` is used you should find anything else that I have forgotten.

**Comment by martin-cs** [comment]
On Fri, 2022-11-18 at 16:03 -0800, Benjamin Valpey wrote:
> 
> > Yes I absolutely agree. I should reiterate that I want the behavior
> > for NaNs to be the same for this function as it is for the to_real,
> > where it is left uninterpreted, rather than deciding on a 
> > representation.

Minor, but perhaps significant, point of pedantry.  fp.to_real is
*completely* underspecified (for technical reasons SMT-LIB refers to
un/underspecified rather than undefined) when the input is NaN.

Do you want fp.to_ieee_bv to be completely underspecified for NaN or to
be guaranteed to return one of the bit patterns that converts to NaN?


> In other words, I don't want my implementation to respond with a
> singular value for NaN. I would like to leave it uninterpreted as it
> is for to_real. For my use case, I would then just define my
> conversion that can handle NaN as follows:
> 
> (define-fun my_fp_ieee_bv32 ((a Float32)) (_ BitVec 32)
>   (ite (fp.isNaN a)
>        (ite (fp.isNegative a) bvNegNan bvPosNaN)
>        (fp.to_ieee_be a)))
>  ; bvPos/NegNaN can be whatever I want, doesn't matter.
> ; fp.to_ieee_bv is the name of the conversion

Again, more pedantry, this may not do what you want as (fp.isNegative
a) is always false when a is NaN.  So is (fp.isPositive).  Note that it
might just be easier to:

(assert (= bvPosNaN (fp.to_ieee_bv NaN)))

as that will constraint the uninterpreted function.


> You will need to at least be able to rewrite total constants ... Note
> that this will fix a particular value for NaN
> 
> I would like to not fix NaN values. I'm away from my machine at the
> moment, but how does fp.to_real handle NaNs, given that they are
> supposed to be left uninterpreted? (Are they?) I would like my
> implementation to do the same thing and not return any value, perhaps
> leaving it as a lambda?

My apologies for being unclear.  I should perhaps explain the different
semantics of the to_real operations.

1. fp.to_real is given by the standard an is underspecified when the
input is an infinity or a NaN.  This has a kind so that parsing and
printing can happen in the expected way.

2. So that all of our functions are total, this is converted super
early on (expandDefinitions) to a TO_REAL_TOTAL.  This is a total
function that takes two arguments.  If the first input is not NaN, then
it converts it and returns it.  Otherwise it returns the second
argument.  In expandDefinitions, we create also create an uninterpreted
function:

floatingpoint_to_real_infinity_and_NaN_case : F_{e,s} -> R 

So the translation is:

(fp.to_real f)
~~~>
(fp.to_real_total f (floatingpoint_to_real_infinity_and_NaN_case f))

That way all of the undefined-ness is separated off into one place; the
uninterpreted function.

3. The constant object FloatingPoint implements two conversions a
"partial" (i.e. returns a pair) convertToRational and a two argument
(one implicit as the object) convertToRationalTotal which have
essentially the same semantics as the two kinds.


What I wanted to communicate was that you can use pack() in cases where
the NaN value should not be taken:

FloatingPoint::PartialBitVector FloatingPoint::convertToIEEEBV() const
{
  return PartialBitVector(pack(), !isNaN());
}

BitVector FloatingPoint::convertToIEEEBVTotal(BitVector nanCase) const
{
  // MUST have EXACTLY the same semantics as the bit-blast
  if (isNaN) {
    // generalise from the single precision case...
    return nanCase | (0xFF << 23) 
                   | (((nanCase & 0x7FFFFF) == 0) ? 1<<22 : 0);
  } else {
    return pack();
  }
}


> > Oh, as a minor aside, you are best off avoiding converting floats 
> > to/from bit-vectors between each instruction if you can help it
> 
> I thought about this, and at least for my use case plan on keeping
> things as floats for as long as possible.

Yes, as general advice for SMT, you should keep "types" as separate as
possible.  The whole design of CDCL(T) / DPLL(T) solvers is based on
the notion that different theory solvers will be basically independent
much of the time.

If you implement the rewrite (to_float (fp.to_ieee_bv f)) ~~> f then it
should avoid a fair section of that overhead.


> For the remainder, I will likely have more questions once I begin
> working out your suggestions. In any case, thanks again for your
> extremely detailed response.

You are welcome; let me know how it goes.

**Reply by bavalpey** [reply]
Hi Martin.

Sorry to necro this old post, but life got in the way. I am now finally revisiting this.

I am still very keen on having support for fp.to_ieee_bv in cvc5. One of the main issues I am facing is that cvc5 blows Z3 out of the water for queries involving floats, but the lack of support for fp.to_ieee_bv makes it quite difficult for my use case (as described).
I think the rewrite rules would solve a lot of this. Not having to issue assertions that are unnecessary simplifies both the model and the steps that the solver has to take.

I have done my best to heed your gracious tips and have forked and wrote a [WIP] commit that implements most of what I think needs to be changed: https://github.com/cvc5/cvc5/compare/cvc5:0f08d9e...bavalpey:8f2432e

I did get stuck, specifically on the rewrite rules.  First, I was quite lost in figuring out exactly _how_ to specify the rewrite for FP -> BV -> FP (likewise BV -> FP -> BV). Not the body per se, just where to put it, what the operands it has, where to have the rewrite be applied, etc.

I also ended up getting lost in writing the convertToIEEEBVTotal. It seems the codebase has changed since your suggestions. While I believe I have correctly used ``pack()`` when the result isn't NaN, it's unclear precisely what should happen when it _is_.  For instance, why are the bits being fixed when you write

```
return nanCase | (0xFF << 23) 
                   | (((nanCase & 0x7FFFFF) == 0) ? 1<<22 : 0);
```

Isn't this now specifying some specific NaN value? Is it not still uninterpreted?



When it came to implementing the bit blasting portion, you were right. It is tricky, and I got quite lost. I can hit my head against the wall on this and try to brute force figure out how to do it, but if you're willing to offer guidance (or more hands on assistance) in terms of exactly how to implement it, it would be greatly appreciated!
